adcc/adc_pp/state2state_transition_dm.py

```diff
@@ -20,12 +20,131 @@
 ## along with adcc. If not, see <http://www.gnu.org/licenses/>.
 ##
 ## ---------------------------------------------------------------------
+from adcc import block as b
 from adcc.AdcMethod import AdcMethod
+from adcc.functions import einsum
 from adcc.AmplitudeVector import AmplitudeVector
+from adcc.OneParticleOperator import OneParticleOperator
 
 import libadcc
 
-DISPATCH = {}  # None implemented
+
+# TODO: just a sketch for what I'd like to have, because we need to check on
+# 2 amplitude vectors simultaneously
+def _check_have_singles_block(*amplitudes):
+    if any("s" not in amplitude.blocks for amplitude in amplitudes):
+        raise ValueError("state2state_transition_dm at ADC(0) level and "
+                         "beyond expects an excitation amplitude with a "
+                         "singles part.")
+
+
+def _check_have_doubles_block(*amplitudes):
+    if any("d" not in amplitude.blocks for amplitude in amplitudes):
+        raise ValueError("state2state_transition_dm at ADC(2) level and "
+                         "beyond expects an excitation amplitude with a "
+                         "singles and a doubles part.")
+
+
+def _check_singles_subspaces(*amplitudes):
+    for amplitude in amplitudes:
+        ul1 = amplitude["s"]
+        if ul1.subspaces != [b.o, b.v]:
+            raise ValueError("Mismatch in subspaces singles part "
+                             f"(== {ul1.subspaces}), where {b.o}{b.v} "
+                             "was expected.")
+
+
+def _check_doubles_subspaces(*amplitudes):
+    for amplitude in amplitudes:
+        ul2 = amplitude["d"]
+        if ul2.subspaces != [b.o, b.o, b.v, b.v]:
+            raise ValueError("Mismatch in subspaces doubles part "
+                             f"(== {ul2.subspaces}), where "
+                             f"{b.o}{b.o}{b.v}{b.v} was expected.")
+
+
+def s2s_tdm_adc0(mp, amplitude_l, amplitude_r, intermediates):
+    _check_have_singles_block(amplitude_l, amplitude_r)
+    _check_singles_subspaces(amplitude_l, amplitude_r)
+    ul1 = amplitude_l["s"]
+    ur1 = amplitude_r["s"]
+
+    dm = OneParticleOperator(mp, is_symmetric=False)
+    dm[b.oo] = -einsum('ja,ia->ij', ul1, ur1)
+    dm[b.vv] = einsum('ia,ib->ab', ul1, ur1)
+    return dm
+
+
+def s2s_tdm_adc2(mp, amplitude_l, amplitude_r, intermediates):
+    _check_have_doubles_block(amplitude_l, amplitude_r)
+    _check_doubles_subspaces(amplitude_l, amplitude_r)
+    dm = s2s_tdm_adc0(mp, amplitude_l, amplitude_r, intermediates)
+
+    ul1 = amplitude_l["s"]
+    ur1 = amplitude_r["s"]
+    ul2 = amplitude_l["d"]
+    ur2 = amplitude_r["d"]
```

Comment:
For these sort of things I tend to prefer
```
ul1, ul2 =  amplitude_l["s"],  amplitude_l["d"]
ur1, ur2 =  amplitude_r["s"],  amplitude_r["d"]
```

but I suppose one can argue about that.

Reply:
Done.